provider/providertest/integration.go

```diff
@@ -0,0 +1,422 @@
+		return fmt.Errorf("%s -> %s", verification.Name, err)
+	}
+
+	if err = compareDataWithExpected(verification.ExpectedValues, data); err != nil {
+		return fmt.Errorf("verification failed for table %s; %s", resource.Table.Name, err)
+	}
+
+	// verify root entry relations
+	for _, e := range data {
+		id, ok := e["id"]
+		if !ok {
+			return fmt.Errorf("failed to get parent id for %s", resource.Table.Name)
+		}
+		if err = verifyRelations(verification.Relations, id, resource.Table.Name, conn); err != nil {
+			return fmt.Errorf("verification failed for relations of table entry %s; id: %v -> %s", resource.Table.Name, id, err)
+		}
+	}
+	return nil
+}
+
+// verifyRelations - recursively runs through all the relations and compares their values with expected data
+func verifyRelations(relations []*ResourceIntegrationVerification, parentId interface{}, parentName string, conn *pgx.Conn) error {
+	for _, relation := range relations {
+		// build query to get relation
+		psql := squirrel.StatementBuilder.PlaceholderFormat(squirrel.Dollar)
+		sq := psql.Select(fmt.Sprintf("json_agg(%s)", relation.Name)).
+			From(relation.Name).
+			LeftJoin(fmt.Sprintf("%[1]s on %[1]s.id = %[3]s.%[2]s", parentName, relation.ForeignKeyName, relation.Name)).
+			Where(squirrel.Eq{fmt.Sprintf("%s.id", parentName): parentId})
+		query, args, err := sq.ToSql()
+		if err != nil {
+			return fmt.Errorf("%s -> %s", relation.Name, err)
+		}
+		row := conn.QueryRow(context.Background(), query, args...)
+		data, err := getDataFromRow(row)
+		if err != nil {
+			return fmt.Errorf("%s -> %s", relation.Name, err)
+		}
+
+		if err = compareDataWithExpected(relation.ExpectedValues, data); err != nil {
+			return fmt.Errorf("%s -> %s", relation.Name, err)
+		}
+
+		// verify relation entry relations
+		for _, e := range data {
+			id, ok := e["id"]
+			if !ok {
+				return fmt.Errorf("failed to get parent id for %s", relation.Name)
+			}
+			if err = verifyRelations(relation.Relations, id, relation.Name, conn); err != nil {
+				return fmt.Errorf("%s id: %v -> %s", relation.Name, id, err)
+			}
+		}
+	}
+	return nil
+}
+
+// compareDataWithExpected - runs through expected values and checks if they are satisfied with received data
+func compareDataWithExpected(expected []ExpectedValue, received []map[string]interface{}) error {
+	var errors []error
+	// clone []map that will be compared
+	toCompare := make([]map[string]interface{}, len(received))
+	for i, row := range received {
+		toCompare[i] = make(map[string]interface{})
+		for key, value := range row {
+			toCompare[i][key] = value
+		}
+	}
+
+	for _, verification := range expected {
+		found := 0
+		for i := 0; i < len(toCompare); i++ {
+			if toCompare[i] == nil {
+				continue // this row is already verified - skip
+			}
+			err := compareData(verification.Data, toCompare[i])
+			if err == nil {
+				toCompare[i] = nil // row passed verification - it won't be used
+				found++
+			} else {
+				errors = append(errors, err)
+			}
```

Comment:
I would invert this if err := compareData(verification.Data, toCompare[i]); err != nil { append + continue } and remove the else